### PR TITLE
WIP: component: prototyping custom flag semantics

### PR DESCRIPTION
**Pull request by sttts**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package kubernetes
+
+// ComponentConfigFooController configures the Foo controller.
+type ComponentConfigFooController struct {
+	F int32
+}
```
Comment (stewart-yu) [inline config.go]: Why the `var` so simple? 
/cc

Comment (stewart-yu) [inline config.go]: Why the `var` so simple?

**Review by stewart-yu — COMMENTED** [review]
/cc

File: staging/src/k8s.io/component-base/cli/flags/example/main.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package example
+
+// Parse parses the command line, potentially calls load and then returns a configuration.
+func Parse(cmdLine []string, load func(fname string, dest *ComponentConfigControllerManager)) (*ComponentConfigControllerManager, error) {
```
Comment (mtaufen) [inline staging/src/k8s.io/component-base/cli/flags/example/main.go]: Ah, ok this is just an example of how one could use the framework? For a second I got confused on why this was coupling parse/load/apply together.

**Comment by mtaufen** [comment]
For context, my variant was roughly this (ignore any minor errors, I haven't tried to compile it).

The main differences are: 
- we pass in a fully-defaulted config up-front
- we don't have to load values into a pre-existing config object prior to apply (because in this variant we have the `target` extraction function)
- users have to cast when they specify the target func, because I used `interface{}` to make it generic

```
// Config is the internal representation of a component config
type Config struct {
	A int32
	B []string
}

// NewConfig returns a Config with default values set
func NewConfig() *Config {
	c := &Config{}
	Default(c)
	return c
}

// Default applies defautls to a Config, if values are not already set
func Default(c *Config) {
	if c.A == 0 {
		c.A = 1
	}
	if c.B == nil {
		c.B = []string{"foo"}
	}
}

// ConfigFlagSet will register an applyFunc every time a flag is registered, and
// can apply all of the applyFuncs to a Config, using values parsed into the tmp Config.
type ConfigFlagSet struct {
	tmp        interface{}
	fs         *pflag.FlagSet
	applyFuncs map[string]func(c *Config)
}

// tmp should already contain the default values
func NewConfigFlagSet(fs *pflag.FlagSet, tmp interface{}) *ConfigFlagSet {
	return &ConfigFlagSet{tmp: tmp, fs: fs}
}

func (fs *ConfigFlagSet) Int32Var(name string, usage string, target func(c interface{}) *int32, merge func(tmp, c int32) int32) {
	ptmp := target(fs.tmp)
	fs.fs.Int32Var(ptmp, name, *ptmp, usage)
	// This direct application may be a reasonable default
	fs.applyFuncs[name] = func(c interface{}) {
		pc := target(c)
		if merge == nil {
			*pc = *ptmp
		} else {
			*pc = merge(*ptmp, *pc)
		}
	}
}

func (fs *ConfigFlagSet) StringSliceVar(name string, usage string,
	target func(c interface{}) *[]string, merge func(tmp, c []string) []string) {
	ptmp := target(fs.tmp)
	fs.fs.StringSliceVar(ptmp, name, *ptmp, usage)
	// This direct application may be a reasonable default
	fs.applyFuncs[name] = func(c interface{}) {
		pc := target(c)
		if merge == nil {
			*pc = []string{}
			for _, v := range *ptmp {
				*pc = append(*pc, v)
			}
		} else {
			// TODO(mtaufen): also consider copy before calling merge here, to avoid side-effects
			*pc = merge(*ptmp, *pc)
		}
	}
}

func (fs *ConfigFlagSet) Apply(c interface{}) {
	for name, apply := range applyFuncs {
		if fs.fs.Changed(name) {
			apply(c)
		}
	}
}

func AddConfigFlags(fs *pflag.FlagSet) func(c interface{}) {
	tmp = NewConfig()
	cfs := NewConfigFlagSet(fs)

	cfs.Int32Var("a-flag", "", func(c interface{}) *int32 {
		return &c.(*Config).A
	}, nil)

	cfs.StringSliceVar("b-flag", "", func(c interface{}) *[]string {
		return &c.(*Config).B
	}, func(a, b []string) []string {
		s := []string{}
		s = append(s, a...)
		s = append(s, b...)
		return s
	})

	return cfs.Apply
}

```

**Comment by justinsb** [comment]
What if we simply had a tag `flag:"foo"` on the fields that we wanted to expose as flags?

**Comment by sttts** [comment]
> What if we simply had a tag `flag:"foo"` on the fields that we wanted to expose as flags?

Many of them do not match 1:1. You also have to define merging logic, custom documentation texts, etc. While we could do that declaratively, we will end up with many tags. Not sure it is worth that complexity.

**Pull request closed without merging**